File: examples/objc-hello-world/hello.m
#import <Foundation/Foundation.h>
#import <Holons/Holons.h>

/// Pure deterministic HelloService.
NSString *HOLGreet(NSString *name) {
  NSString *n = (name.length == 0) ? @"World" : name;
  return [NSString stringWithFormat:@"Hello, %@!", n];
}

#ifndef TEST_BUILD
int main(int argc, const char *argv[]) {
  @autoreleasepool {
    if (argc > 1 && strcmp(argv[1], "serve") == 0) {
      NSMutableArray<NSString *> *serveArgs = [NSMutableArray array];
      for (int i = 2; i < argc; i++) {
        [serveArgs addObject:[NSString stringWithUTF8String:argv[i]]];
      }

      NSString *listenURI = HOLParseFlags(serveArgs);
      fprintf(stderr, "objc-hello-world listening on %s\n", listenURI.UTF8String);
      printf("{\"message\":\"%s\"}\n", HOLGreet(@"").UTF8String);
      return 0;
    }

    NSString *name = argc > 1 ? [NSString stringWithUTF8String:argv[1]] : @"";
    printf("%s\n", HOLGreet(name).UTF8String);
    return 0;
  }
}
#endif

File: examples/hello-world/kotlin-hello-world/src/main/kotlin/org/organicprogramming/hello/Hello.kt
package org.organicprogramming.hello

import org.organicprogramming.holons.Serve

/** Pure deterministic HelloService. */
object HelloService {
    fun greet(name: String): String {
        val n = name.ifEmpty { "World" }
        return "Hello, $n!"
    }
}

fun main(args: Array<String>) {
    if (args.firstOrNull() == "serve") {
        val listenURI = Serve.parseFlags(args.drop(1).toTypedArray())
        System.err.println("kotlin-hello-world listening on $listenURI")
        println("""{"message":"${HelloService.greet("")}"}""")
        return
    }

    val name = args.firstOrNull() ?: ""
    println(HelloService.greet(name))
}
